### ISSUE TITLE
How can I ensure that exactly one argument has been supplied?

### ISSUE THREAD

**Issue by roxspring** [tj/commander.js#1000]
I've tried defining a command with a single required argument, and attempted to validate that exactly one source specified:

```
import * as program from 'commander';
import { version } from 'pjson';

program
  .version(version)
  .arguments('<source.md>')
  .action((source: string): void => {
    console.log('source=' + source);
    program.source = source;
  })
  .parse(process.argv);

console.log(program.args);
console.log(program.args.length);

if (program.args.length !== 1) {
  console.log('Invalid arguments');
  program.help();
}

console.log(program.source);
```

And running the program with no arguments, correctly prints 'Invalid arguments' since `program.args.length` is 0. But with 1 argument (`program a`) `program.args.length` is 2!? and outputs the following, claiming that `program.args.length` is 2!?

```
source=a
[
  'a',
  Command {
    ...
  }
]
2
Invalid arguments
```

Weirdly running with four arguments `program a b c d` also appears to drop the `b` argument somehow:

```
source=a
[
  'a',
  Command {
    ...
  },
  'c',
  'd'
]
4
Invalid arguments
```

So I guess I have two questions:

1. Where's the second argument gone?
2. (more importantly) how can I ensure that exactly one argument has been supplied?

**Comment by shadowspawn** [comment]
There is not currently error detection for excess arguments, and it is not currently easy to handle it yourself in a general way. In simple cases you can check the argument count or the original arguments (`process.argv` or `program.rawArgs`).

Open issue is #749 and has some other ideas.

**Comment by shadowspawn** [comment]
There are improvements on the develop branch for the next release (#1030 and #1048).

**Comment by shadowspawn** [comment]
v4.0.0-0 prerelease published: #1067

**Comment by shadowspawn** [comment]
v4.0.0 has been released with a fix for `program.args` which was getting the command appended (matching the arguments passed to the action handler).

Hopefully more useful now!

Feel free to open a new issue if still problems, with new information and renewed interest.
